lib/manager.js

```diff
@@ -0,0 +1,101 @@
+"use strict";
+
+let { bindMethodContext } = require("./util");
+let { createFile, generateFingerprint, uriJoin } = require("faucet-pipeline-util");
+let path = require("path");
+
+module.exports = class AssetManager {
+	constructor(referenceDir, { manifest, fingerprint, exitOnError } = {}) {
+		this.referenceDir = referenceDir;
+		this.manifest = manifest;
+		this.fingerprint = fingerprint;
+		this.exitOnError = exitOnError;
+		this._index = {};
+
+		// bind methods for convenience
+		bindMethodContext(this, "writeFile", "resolvePath");
+	}
+
+	writeFile(filepath, data, error) {
+		let originalPath = filepath;
+		if(this.fingerprint) {
+			filepath = generateFingerprint(filepath, data);
+		}
+
+		return createFile(filepath, data).
+			then(_ => this._updateManifest(originalPath, filepath)).
+			then(_ => {
+				this._report(originalPath, error);
+				if(error && this.exitOnError) {
+					abort();
+				}
+			}).
+			catch(abort);
+	}
+
+	resolvePath(filepath, { nodeResolution } = {}) {
+		if(filepath.substr(0, 2) === "./") {
+			return path.resolve(this.referenceDir, filepath);
+		} else if(!nodeResolution) {
+			throw new Error(`path must be relative: \`${repr(filepath)}\``);
+		} else { // attempt via Node resolution algorithm
+			try {
+				return resolveModulePath(filepath, this.referenceDir);
+			} catch(err) {
+				throw new Error(`could not resolve \`${repr(filepath)}\``);
+			}
+		}
+	}
+
+	_updateManifest(originalPath, actualPath) {
+		[originalPath, actualPath] = [originalPath, actualPath].map(filepath => {
+			return path.relative(this.referenceDir, filepath);
+		});
+
+		let { file: filepath, baseURI } = this.manifest;
+		/* eslint-disable indent */
+		this._index[originalPath] = baseURI.call ?
+				baseURI(actualPath, path.basename(actualPath)) :
+				uriJoin(baseURI, actualPath);
+		/* eslint-enable indent */
+		return createFile(filepath, JSON.stringify(this._index) + "\n");
```

Comment:
What if the user doesn't want to have a manifest file? This step has to be optional IMHO

Reply:
Looks like I'd forgotten about that. I suppose this method should just abort immediately if `this.manifest === false`?

Reply:
No, it should still update the manifest – it should just not write the file. The reason is the interplay between sass and static: sass can access the in-memory manifest 😸

Reply:
In that case, we should remove the underscore from `this._index`?